Implement BTB using the cache library

pranith opened the pull request:
This enables the BTB to be associative and use various replacement policies.

pranith committed:
cpu: pred: Use cache library to build the BTB

This allows us to make the BTB associative and also experiment with various
replacement and indexing policies for the BTB.

Tested with setting associativity to 1 to reflect current behavior.

Change-Id: Ie9085fc4728e000a9b62ed48a0ecd1a57a22201c

pranith commented:
@giactra I tried using the templatized version of the cache library as an exercise. Can you please review this change? Thank you!

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+    using Params = BaseIndexingPolicyParams;
+};
+
+class BTBEntry : public CacheEntry
```
I wouldn't inherit from CacheEntry, as this is only adding complexity IMHO with no gain. It forces you to overload and hide some methods. The BTB entry should inherit from ReplaceableEntry

pranith replied:
I tried inheriting it from ReplaceableEntry and that just seems to be doing most of what CacheEntry is already doing for valid and other methods. We are duplicating most of the code here again.

Is there any reason why you prefer inheriting from ReplaceableEntry instead of CacheEntry and avoid duplication?

giactra replied:
There are several reasons IMO

1) Design concern: Public virtual methods define the interface of a class. If you inherit from a class but you end up overloading such interface, then you are basically saying the BtbEntry is not really a CacheEntry, it is just reusing some of its code.

2) Possible bugs: this hiding can lead to potential bugs, see https://github.com/gem5/gem5/pull/1374

3) Tag ambiguity: the idea of defining a tag as the single address is just an artifact of past design. In this case the tag type is really the pair of addr and tid. But by using the CacheEntry you are basically defining two different tag types.

4) Performance: most of the BTBEntry logic is calling virtual functions. So you are needlessly adding entries to a VTABLE for a polymorphism that is never used, and you are calling them paying the vtable indirection penalty.

I understand that some code will look the same between the CacheEntry and the BTBEntry but IMHO the overhead is minimal and we would benefit a lot in decoupling the two entries, simplifying the design.

pranith replied:
Thanks for the explanation @giactra. BtbEntry is derived from CacheEntry means it has some properties of CacheEntry, but not all. This is I think a textbook example of OOP (mammals -> humans).

But, I am not really strong against implementing it from scratch. Just that, there will be code duplication.

I pushed a patch addressing all the review comments. Please take a look!

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+    /**
+     * Convenience typedef.
+     */
+    typedef BTBIndexingPolicyParams Params;
```
Nit: prefer using for consistency for what you are doing above with KeyType

pranith replied:
Done.

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+};
+
+
+class BTBIndexingPolicy : public SetAssociative
```
Similarly to the BTBEntry discussion above. We need to pay the price and not inherit from SetAssociative. SetAssociative and the parent BaseIndexingPolicy are assuming an Addr key, whereas here our key is the Addr + Tid pair. I suggest you to have a look at what I did with TaggedIndexingPolicy... This is the hard bit I recon, as it requires you to define a templated SimObject. Happy to provide further assistance if needed

pranith replied:
@giactra I tried your suggestion and created a BTBIndexingPolicy/BTBSetAssociative policy derived from SimObject. However, I seem to be running into a linking error. Can you please help?

pranith replied:
Ah, looks like I missed implementing the regenerateAddr method. It works now.

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+  private:
+    unsigned log2NumThreads;
+
+    /* Mark these private as we need tid  */
+    using SetAssociative::extractSet;
+    using SetAssociative::getPossibleEntries;
```
Ditto this is a workaround

pranith replied:
Removed.

giactra commented on src/cpu/pred/simple_btb.cc:
```diff
@@ -44,84 +44,37 @@
-
-    assert(btb_idx < numEntries);
-
-    if (btb[btb_idx].valid
-        && inst_tag == btb[btb_idx].tag
-        && btb[btb_idx].tid == tid) {
-        return &btb[btb_idx];
-    }
```
With your change we are not checking if the btb entry is valid anymore. You need to check for validity in the match method

pranith replied:
The CacheEntry's 'match' checks for the valid flag.

giactra replied:
Ah right

giactra reviewed (changes requested):
Thanks @pranith for this change! As this is a purely refactoring change, could you run some workload with O3 and see if there are any stats difference between pre and after?

pranith commented:
@giactra Thanks for the review! I tried inheriting from the base classes (instead of CacheEntry and IndexingPolicy) but it seems like we need to copy most of what these classes provide. I feel it becomes needlessly more labor intensive and error prone to copy the code and duplicate it in multiple places. 

Can you please provide the rationale for using the base classes? Only reason I can think of is that we are overriding base class methods with methods of a different signature, but I don't see what is wrong in doing this.

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+     * @return The possible entries.
+     */
+    std::vector<ReplaceableEntry*>
+    getPossibleEntries(const KeyType &key) const
```
Please mark this as override

pranith replied:
Done.

giactra commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -60,77 +84,31 @@ class SimpleBTB : public BranchTargetBuffer
      *  @param inst_PC The branch to look up.
      *  @return Returns the index into the BTB.
      */
     inline unsigned getIndex(Addr instPC, ThreadID tid);
```
It looks like this has been removed

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -60,77 +84,31 @@ class SimpleBTB : public BranchTargetBuffer
-} // namespace gem5
+} // namespace gem5::branch_prediction
+
+}
```
nit: re-add `// namespace gem5`

pranith replied:
Done.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -41,17 +41,41 @@
 namespace gem5
 {
 
+using BTBIndexingPolicy = IndexingPolicyTemplate<branch_prediction::BTBTagTypes>;
```
I think this line exceeds the character limit (Lines must be a maximum of 79 characters long).

pranith replied:
Fixed.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -41,17 +41,41 @@
 {
 
+/**
+ * This helper generates an a tag extractor function object
```
nit: ´an a´

pranith replied:
Fixed.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -41,17 +41,41 @@
 #ifndef __CPU_PRED_SIMPLE_BTB_HH__
 #define __CPU_PRED_SIMPLE_BTB_HH__
 
+#include <memory>
```
<memory> is not used in this file

pranith replied:
Removed.

odanrc commented on src/cpu/pred/simple_btb.cc:
```diff
@@ -44,84 +44,37 @@
-        instShiftAmt(p.instShiftAmt),
-        log2NumThreads(floorLog2(p.numThreads))
+      btb("simpleBTB", p.numEntries, p.associativity,
+          p.btbReplPolicy, p.btbIndexingPolicy, BTBEntry(genTagExtractor(p.btbIndexingPolicy)))
```
Ditto line length

pranith replied:
Fixed.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+    {}
+
+    /** Update the entry in the BTB */
+    void
```
I think this function deserves an improvement in the documentation. Something to explain its difference from the insert() function. It seems to be a workaround due to the fact that the insertion only uses the key and the target and inst are not part of the key.

pranith replied:
Done.

odanrc replied:
I don´t see the change. Have you pushed it?

pranith replied:
Sorry! missed pushing it earlier... pushed it in the latest commit.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,207 @@
+
+    /** Update the entry in the BTB */
+    void
+    update(ThreadID _tid,
```
Since it seems that we only update() right after calling insert, the _tid assignment is redundant and should probably be dropped.

pranith replied:
Removed tid argument.

odanrc commented on configs/common/cores/arm/HPI.py:
```diff
@@ -1683,6 +1683,11 @@ class HPI_MMU(ArmMMU):
+    associativity = 1
+    btbReplPolicy = LRURP()
+    btbIndexingPolicy = BTBIndexingPolicy(
+        size="384B", entry_size=3, assoc=associativity
```
Curiosity: where do the config numbers come from? (I am not too familiar with BTBs)

pranith replied:
It is approximately 'numEntires * entry_size' so that indexing is as before. I will re-work this part once before merging the final changes.

pranith commented:
> Thanks @pranith for this change! As this is a purely refactoring change, could you run some workload with O3 and see if there are any stats difference between pre and after?

I reran with the patch and there were no stat differences in BTB. I updated the BTB params to keep the same behavior as before.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    std::string
+    print() const override
+    {
+        return csprintf("tag: %#x valid: %d | %s", getTag(),
```
The tid should also be printed.

pranith replied:
Updated.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    }
+
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other)
```
The rule of 3 dictates that if we need to define copy constructor then we should also need to define the copy assignment operator.

pranith replied:
Added a copy assignment operator.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * Get tag associated to this block.
+     */
+    virtual Addr getTag() const { return tag; }
```
I think that it should instead return KeyType, since the tid *is* part of the tag. We need to make sure that what happened with the prefetchers (some use the secure bit, and others don´t) doesn´t happen in this class. We can already see an instance of the bug happening in this class´ print() and invalidate(), where the tid was forgotten.

pranith replied:
Changed the tag type to be KeyType througout.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    bool valid;
+
+    /** The entry's tag. */
+    Addr tag;
```
Instead of storing it as an address, we should store it as a KeyType. This class should not use Addr.

pranith replied:
Done.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    StaticInstPtr inst;
+
+    /** The entry's thread id. */
+    ThreadID tid;
```
As per the other comments, this should be the same visibility as the ´tag´ variable.

pranith replied:
Removed the independent tid member.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * Set tag associated to this block.
+     */
+    virtual void setTag(Addr _tag) { tag = _tag; }
```
It should have used KeyType, because otherwise we end up forgetting about tid (see invalidate() using this function and not assigning a ´max tid´).

pranith replied:
Done.

pranith commented:
@odanrc Thaks for the review. I squashed all the commits into one. Not sure if this is can be done during merge?

powerjg commented:
> I reran with the patch and there were no stat differences in BTB. I updated the BTB params to keep the same behavior as before.

I just want to double check that even though you saw there were no stats differences, did you make sure the BTB is being used in the tests? Our test coverage of many models like the BTB is pretty bad. You would likely have had to write your own test for this.

pranith commented:
> 
> I just want to double check that even though you saw there were no stats differences, did you make sure the BTB is being used in the tests? Our test coverage of many models like the BTB is pretty bad. You would likely have had to write your own test for this.

The BTB sees the same number of accesses, hits, and, misses with the change for a small test program. So yes, I can confirm that it is being exercised and behaves as before the patch.

odanrc commented:
> @odanrc Thaks for the review. I squashed all the commits into one. Not sure if this is can be done during merge?

I think for this small PR there will be no issue, but since there are other reviewers still reviewing I think it would have been better to wait for all approvals to be in order, in case someone was using the commits as part of their review methodology. IIRC, the merge does not trigger a merge squash, but I may be wrong.

pranith commented:
@powerjg Can you please point me to docs on how to run the CI tests locally? It seems some quick boot tests are failing.

powerjg commented:
See https://github.com/gem5/gem5/blob/stable/TESTING.md#running-tests-in-batch

pranith commented:
@giactra Can you please review? This now passes CI tests and is ready for merge. Please let me know if you want me to squash the two commits together or you can do it during merge. Thanks!

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+    /**
+     * Checks if the entry is valid.
+     */
+    virtual bool isValid() const { return valid; }
+
+    /**
+     * Get tag associated to this block.
+     */
+    virtual KeyType getTag() const { return tag; }
+
+    /** Invalidate the block. Its contents are no longer valid. */
+    virtual void
+    invalidate()
+    {
+        valid = false;
+        setTag({MaxAddr, -1});
+    }
```
Are you planning on extending this class? Otherwise I would consider devirtualizing these methods

pranith replied:
Updated.

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+    }
+
+  protected:
+    /**
+     * Set tag associated to this block.
+     */
+    virtual void setTag(KeyType _tag) { tag = _tag; }
+
+    /** Set valid bit. The block must be invalid beforehand. */
+    virtual void
+    setValid()
+    {
+        assert(!isValid());
+        valid = true;
+    }
```
Ditto

pranith replied:
Removed virtual.

giactra commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+    {
+        auto set_idx = extractSet(key);
+
+        fatal_if(set_idx >= std::pow(2, tagShift - setShift),
+            "Set index outside range");
```
This looks expensive. can't we just assert (meaning we can skip this in the fast build) and simply check if set_idx is bigger or equal than sets.size()

This is what we were doing in the old version:

`    assert(btb_idx < numEntries);`

pranith replied:
Updated as suggested.

giactra commented on configs/common/cores/arm/HPI.py:
```diff
@@ -1683,6 +1683,11 @@ class HPI_MMU(ArmMMU):
+    associativity = 1
+    btbReplPolicy = LRURP()
+    btbIndexingPolicy = BTBSetAssociative(
+        size="512B", entry_size=4, assoc=associativity
```
I feel your pain here. I understand you are carefully choosing size and entry_size to account for numEntries and the instShift.

It is 100% not your fault but I believe this is a bit fragile. What happens if someone changes numEntries in the BTB?

To overcome this problem I just created:

https://github.com/gem5/gem5/pull/1557

It will allow you to do

    btbIndexingPolicy = BTBSetAssociative(
        num_entries=numEntries, set_shift=instShiftAmt, assoc=associativity)

Would you be up for waiting for my PR to be merged?

pranith replied:
Thanks for the PR @giactra. It definitely makes it much easier to instantiate the structures. I will rebase on top of your pull request and update this patch.

pranith replied:
Updated accordingly.

giactra reviewed (commented):
Hi @pranith I have some minor comments and a more important one that relates to my new PR: https://github.com/gem5/gem5/pull/1557

(I developed the PR once I saw this one here and its problems)

pranith commented:
@andysan Now that we have the tag based addressing in the cache, can you please review the 'tagMask' commit of this PR? We need tagMask for the BTB (which I just realized).Thanks!

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+
+namespace gem5 {
+
+class BTBTagTypes
```
Nit: this is a tag type - i.e., not plural

pranith replied:
Updated.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+    PARAMS(BTBSetAssociative);
+    using KeyType = BTBTagTypes::KeyType;
+
+    BTBSetAssociative(const Params &p) : BTBIndexingPolicy(p, p.num_entries, p.set_shift)
```
Fix to fit in line length

pranith replied:
Fixed.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+    Addr regenerateAddr(const KeyType &key,
+                        const ReplaceableEntry* entry) const override
+    {
+        return (key.address << tagShift) | (entry->getSet() << setShift);
```
Is this function being called and its output verified anywhere? At a glance this does not seem to be correct because it lacks the xor on the tid

pranith replied:
I don't think we use this function. I replaced the body with an 'assert(false)'. It would be good to remove it from the base class and only define it for mem-cache related policies, but it didn't seem straight forward.

odanrc replied:
Regarding not having it, if you are sure it is not going to be needed, then sure we don´t need to implement it. One alternative is to make the base class no longer pure virtual and panic there. Speaking of which, it would be better to use something in the lines of ´panic(not implemented)´ than the assertion, as it is not a user error, but a lack of feature.

pranith replied:
I don't know what to provide in the default non-pure virtual function, so I just updated the assert() to a panic.

odanrc commented on src/cpu/pred/btb_entry.hh:
```diff
@@ -0,0 +1,275 @@
+     * Checks if the given tag information corresponds to this entry's.
+     */
+    bool
+    match(const KeyType &key)
```
This should probably be a const function - I don´t think we ever expect to change the class when checking for a match

pranith replied:
Updated.

andysan commented:
> @andysan Now that we have the tag based addressing in the cache, can you please review the 'tagMask' commit of this PR? We need tagMask for the BTB (which I just realized).Thanks!

My main concern with the tag bits param is that there is no way for users to reason about this. Does setting it to something other than 64 truncate the TID or the address? What if we add even more information (we'll need to do this if we want to be representative of real hardware) in the tags does it truncate the same fields as before? I think realistically, we can't expect users to be able to use this param.

A better solution would be to deal with truncation in the btb indexing policy instead where the semantic information is available. That'd let users decide how many bits of address to include, whether to include the TID, VMID, etc., which would be much more meaningful than just truncating some bits.

I'm currently out of office, so I won't have time to look at this in more detail. @giactra could you have a look and drive this from our side?

pranith commented:
> A better solution would be to deal with truncation in the btb indexing policy instead where the semantic information is available. That'd let users decide how many bits of address to include, whether to include the TID, VMID, etc., which would be much more meaningful than just truncating some bits.

I updated BTB indexing policy to use tag mask independent of the base policy. There are other users of cache library that need similar tag mask like bop prefetcher, but I left it alone for now.

> 
> I'm currently out of office, so I won't have time to look at this in more detail. @giactra could you have a look and drive this from our side?

@giactra Can you please review? Thanks!

pranith commented:
@giactra @odanrc Can you please review this updated patch? It would be good to have it merged if there are no new comments.

odanrc commented on src/cpu/pred/BranchPredictor.py:
```diff
@@ -83,6 +85,38 @@ class BranchTargetBuffer(ClockedObject):
+    abstract = True
+    cxx_class = "gem5::IndexingPolicyTemplate<gem5::BTBTagType>"
+    cxx_header = "cpu/pred/btb_entry.hh"
+    cxx_template_params = ["class Types"]
```
How does this line work? Is this string a description of the template param?

pranith replied:
I copy+pasted this from the 'BaseIndexingPolicy' from 'mem/cache/tags/indexing_policies/IndexingPolicies.py' since I am trying to reuse the IndexingPolicyTemplate just like it does. Please let me know if this needs to be updated.

odanrc commented:
@pranith I cannot spot the differences between the previous version and the new version. This is probably because the commits were squashed together. Could you point to where we should focus the review efforts? As a side note, in the future I recommend avoiding squashing new commits added to a change that was already under review, as it makes spotting the changes much harder.

pranith commented:
> @pranith I cannot spot the differences between the previous version and the new version. This is probably because the commits were squashed together. Could you point to where we should focus the review efforts? As a side note, in the future I recommend avoiding squashing new commits added to a change that was already under review, as it makes spotting the changes much harder.

I apologize for this. There were some conflicts in the merge and I had to rebase to fix the conflicts. I will add commits addressing review commits only in the future.

The main change from your last review is the addition of the 'tagMask' field in the BTBIndexingPolicy. Rest of it remains the same.

giactra commented on src/cpu/pred/BranchPredictor.py:
```diff
@@ -93,6 +127,20 @@ class SimpleBTB(BranchTargetBuffer):
         Parent.instShiftAmt, "Number of bits to shift instructions by"
     )
+    associativity = Param.Unsigned(1, "BTB associativity")
+    size = Param.MemorySize("16384", "Size of BTB")
```
Is this a leftover from the initial version?

pranith replied:
Indeed, removed the line. I will squash the commits if you think this is OK. Thanks!

giactra replied:
Yes feel free to squash

pranith replied:
Done.

giactra reviewed (changes requested):
Hi @pranith apologies for the late review. I only have one comment left. Then if @odanrc is also happy I will approve the PR
